g2p/transducer/__init__.py

```diff
@@ -128,30 +129,159 @@ def debugger(self, value):
-        return self
+        return [self]
 
     @tiers.setter
     def tiers(self, value):
         raise ValueError(
             f"Sorry, you tried to change the tiers to {value} but they cannot be changed"
         )
 
-    def pretty_edges(self):
-        edges = copy.deepcopy(self._edges)
+    def pretty_edges(self) -> List[List[Tuple[str, str]]]:
+        """List of edges for each tier (there is only one tier), expressed as strings"""
+        edges = self._edges[:]
         edges.sort(key=lambda x: x[0])
-        for i, edge in enumerate(edges):
-            if edge[1] is None:
-                edges[i] = [self._input_nodes[edge[0]][1], None]
+        out_edges = []
+        for edge in edges:
+            if edge[1] is None:  # FIXME: This should not happen???
+                out_edges.append((self._input_nodes[edge[0]][1], None))
             else:
-                edges[i] = [
-                    self._input_nodes[edge[0]][1],
-                    self._output_nodes[edge[1]][1],
-                ]
-        return edges
+                out_edges.append(
+                    (
+                        self._input_nodes[edge[0]][1],
+                        self._output_nodes[edge[1]][1],
+                    )
+                )
+        return [out_edges]
 
-    def as_dict(self):
+    def alignments(self, edges=None) -> List[Tuple[str, str]]:  # noqa: C901
+        """Alignments of input to output substrings for this graph.
+
+        As opposed to `pretty_edges`, this method will return the
+        one-to-many or many-to-one alignments produced by a
+        conversion.  For example, assuming this input and output with
+        these edges::
+
+            ABCDEFF
+            aabbcdef
+            [(0, 0), (0, 1), (1, 2), (1, 3), (2, 4), (3, 5),
+             (4, 6), (5, 7), (6, 7)]
+
+        it should return::
+
+            [('A', 'aa'), ('B', 'bb'), ('C', c'), ('D', 'd'),
+             ('E', 'e'), ('FF', 'f')]
+
+        In the case of reorderings, the substrings returned correspond
+        to the minimal *monotonic* alignments, that is, the minimal
+        subsequences which preserve ordering between the input and
+        output.  So, for example given these edges::
+
+            ABCDEF
+            abefcd
+            [(0, 0), (1, 1), (2, 4), (3, 5), (4, 2), (5, 3)]
+
+        it should return::
+
+            [('A', 'a'), ('B', 'b'), ('CDEF', 'efcd')]
+
+        This means, that in the case of truly radical reorderings,
+        there is no "alignment" per se.  This is highly unlikely to
+        occur, given the constrained nature of g2p rules.
+
+        Note that in the case of multi-character outputs such as in
+        ARPABET, the output substrings *do not necessarily* correspond
+        to tokens, and spaces are treated as any other character in
+        the alignment (thus, alignments of (' ', ' ') are possible).
+        If you wish to reconstruct the input it is your resonsibility
```

Comment:
great comment! I would prefer to see the `alignments` method renamed though to something that reflects this more, especially since monotonicity is not a requirement of `g2p`. `monotonic_alignments` seems too long though...suggestions?

Reply:
I agree, `alignments` could mean anything, actually, I would prefer to rename `edges` to `alignments` to clarify that it's now always the actual input-to-output alignments for the composed graph, and put `edges` back to its original semantics.

Since the most important thing about `alignments` is that it returns a sequence of aligned substrings, perhaps it should be called `aligned_substrings`?